README.rst
jpredapi
========

.. image:: https://img.shields.io/pypi/l/jpredapi.svg
   :target: https://choosealicense.com/licenses/mit
   :alt: License information

.. image:: https://img.shields.io/pypi/v/jpredapi.svg
   :target: https://pypi.org/project/jpredapi
   :alt: Current library version

.. image:: https://img.shields.io/pypi/pyversions/jpredapi.svg
   :target: https://pypi.org/project/jpredapi
   :alt: Supported Python versions

.. image:: https://readthedocs.org/projects/jpredapi/badge/?version=latest
   :target: https://jpredapi.readthedocs.io/en/latest/?badge=latest
   :alt: Documentation status

.. image:: https://api.travis-ci.org/MoseleyBioinformaticsLab/jpredapi.svg?branch=master
   :target: https://travis-ci.org/MoseleyBioinformaticsLab/jpredapi
   :alt: Travis CI status


The `jpredapi` package provides a simple Python interface for submitting
and retrieving jobs from JPred: A Protein Secondary Structure Prediction Server
(JPred_).

This is unofficial Python port of `jpredapi` perl script that can be found here:
http://www.compbio.dundee.ac.uk/jpred4/api.shtml


Links
~~~~~

   * jpredapi @ GitHub_
   * jpredapi @ PyPI_
   * Documentation @ ReadTheDocs_


Installation
~~~~~~~~~~~~

Install on Linux, Mac OS X
--------------------------

.. code:: bash

   python3 -m pip install jpredapi

Install on Windows
------------------

.. code:: bash

   py -3 -m pip install jpredapi

.. note:: Read the `User Guide`_ and `The jpredapi Tutorial`_ on ReadTheDocs_
          to learn more and to see code examples on using the `jpredapi` as a
          library and as a command-line tool.


License
~~~~~~~

This package is distributed under the MIT_ `license`.


.. _pip: https://pip.pypa.io
.. _JPred: http://www.compbio.dundee.ac.uk/jpred

.. _GitHub: https://github.com/MoseleyBioinformaticsLab/jpredapi
.. _ReadTheDocs: http://jpredapi.readthedocs.io
.. _PyPI: https://pypi.org/project/jpredapi
.. _User Guide: http://jpredapi.readthedocs.io/en/latest/guide.html
.. _The jpredapi Tutorial: http://jpredapi.readthedocs.io/en/latest/tutorial.html
.. _MIT: https://choosealicense.com/licenses/mit


CHANGELOG.rst
.. :changelog:

Release History
===============

1.5.6 (2018-07-20)
~~~~~~~~~~~~~~~~~~

**Improvements**

- Fixed license badge display.
- Updated links.


1.5.4 - 1.5.5 (2018-07-19)
~~~~~~~~~~~~~~~~~~~~~~~~~~

**Improvements**

- Changed license to MIT.


1.5.3 (2018-07-11)
~~~~~~~~~~~~~~~~~~

**Improvements**

- API calls return actual `Response` object instead of returning `None`.
- Added "--attempts=<max>" option to specify maximum number of attempts before giving up.
- Added "check_rest_version" command.
- Added "mock" and "real" tests.

**Bugfixes**

- Added check that job was actually created on JPred server.


1.5.2 (2018-06-19)
~~~~~~~~~~~~~~~~~~

**Improvements**

- Isolated CLI into its own module.
- Added entry point, i.e. can use `$ jpredapi` instead of `python3 -m jpredapi`.


1.5.1 (2017-11-29)
~~~~~~~~~~~~~~~~~~

**Bugfixes**

- Minor bug fixes and cleanup.


1.5.0 (2017-02-07)
~~~~~~~~~~~~~~~~~~

- Initial public release 
  (version "1.5.0" matches the JPred REST API version).


tests/test_as_cli.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import subprocess
import pytest


SKIP_REAL_JPREDAPI = bool(int(os.getenv("SKIP_REAL_JPREDAPI", True)))


@pytest.mark.skipif(SKIP_REAL_JPREDAPI, reason="Skipping tests that hit the real JPred API server.")
@pytest.mark.parametrize("cmdargs", [
    "--help",
    "--version",
    "check_rest_version",
    "check_rest_version --silent",
    "quota --email=name@domain.com",
    "quota --email=name@domain.com --silent",
    "submit --mode=single --format=raw --seq=MQVWPIEGIKKFETLSYLPP",
    "submit --mode=single --format=raw --file=tests/example_data/single_raw.example",
    "submit --mode=single --format=fasta --file=tests/example_data/single_fasta.example",
    "submit --mode=batch --format=fasta --file=tests/example_data/batch_fasta.example --email=name@domain.com",
    "submit --mode=msa --format=fasta --file=tests/example_data/msa_fasta.example --email=name@domain.com",
    "submit --mode=msa --format=msf --file=tests/example_data/msa_msf.example --email=name@domain.com",
    "submit --mode=msa --format=blc --file=tests/example_data/msa_blc.example --email=name@domain.com",
    "status --jobid=jp_K46D05A",
    "status --jobid=jp_K46D05A --results=jpred_sspred/results",
    "status --jobid=jp_K46D05A --results=jpred_sspred/results --extract",
    "get_results --jobid=jp_K46D05A",
    "get_results --jobid=jp_K46D05A --results=jpred_sspred/results",
    "get_results --jobid=jp_K46D05A --results=jpred_sspred/results --extract"
])
def test_cli(cmdargs):
    cmdargs = ["python", "-m", "jpredapi"] + cmdargs.split()
    subprocess.check_output(cmdargs)


tests/test_as_library.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import re
import os
import pytest
import jpredapi

try:
    from unittest.mock import patch
except ImportError:
    from mock import patch


SKIP_REAL_JPREDAPI = bool(int(os.getenv("SKIP_REAL_JPREDAPI", True)))


def test_check_rest_version():
    version = "v.1.5"
    with patch("jpredapi.check_rest_version") as mock_version:
        mock_version.return_value.version = "v.1.5"
    assert mock_version.return_value.version == version


@pytest.mark.skipif(SKIP_REAL_JPREDAPI, reason="Skipping tests that hit the real JPred API server.")
def test_check_rest_version_real():
    version = jpredapi.check_rest_version()
    assert version == 'v.1.5'


def test_quota():
    with patch('jpredapi.quota') as response:
        response.return_value.status_code = 200
    assert response.return_value.status_code == 200


@pytest.mark.skipif(SKIP_REAL_JPREDAPI, reason="Skipping tests that hit the real JPred API server.")
def test_quota_real():
    response = jpredapi.quota(email="name@domain.com")
    assert response.status_code == 200


@pytest.mark.parametrize("mode,user_format,file,seq,skipPDB,email,name,silent,host", [
    ("single", "raw", None, "MQVWPIEGIKKFETLSYLPP", True, None, None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("single", "raw", "tests/example_data/single_raw.example", None, True, None, None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("single", "fasta", "tests/example_data/single_fasta.example", None, True, None, None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("batch", "fasta", "tests/example_data/batch_fasta.example", None, True, "name@domain.com", None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("msa", "fasta", "tests/example_data/msa_fasta.example", None, True, "name@domain.com", None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("msa", "msf", "tests/example_data/msa_msf.example", None, True, "name@domain.com", None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("msa", "blc", "tests/example_data/msa_blc.example", None, True, "name@domain.com", None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest")
])
def test_submit(mode, user_format, file, seq, skipPDB, email, name, silent, host):
    with patch('jpredapi.submit') as mock_submit:
        mock_submit.return_value.status_code = 202
        mock_submit.return_value.text = "Created JPred job"

        response = jpredapi.submit(mode=mode, user_format=user_format, file=file, seq=seq, skipPDB=skipPDB,
                                   email=email, name=name, silent=silent, host=host)
        assert response.status_code == 202 and "Created JPred job" or "You have successfully submitted" in response.text


@pytest.mark.skipif(SKIP_REAL_JPREDAPI, reason="Skipping tests that hit the real JPred API server.")
@pytest.mark.parametrize("mode,user_format,file,seq,skipPDB,email,name,silent,host", [
    ("single", "raw", None, "MQVWPIEGIKKFETLSYLPP", True, None, None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("single", "raw", "tests/example_data/single_raw.example", None, True, None, None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("single", "fasta", "tests/example_data/single_fasta.example", None, True, None, None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("batch", "fasta", "tests/example_data/batch_fasta.example", None, True, "name@domain.com", None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("msa", "fasta", "tests/example_data/msa_fasta.example", None, True, "name@domain.com", None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("msa", "msf", "tests/example_data/msa_msf.example", None, True, "name@domain.com", None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"),
    ("msa", "blc", "tests/example_data/msa_blc.example", None, True, "name@domain.com", None, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest")
])
def test_submit_real(mode, user_format, file, seq, skipPDB, email, name, silent, host):
    response = jpredapi.submit(mode=mode, user_format=user_format, file=file, seq=seq, skipPDB=skipPDB,
                               email=email, name=name, silent=silent, host=host)
    assert response.status_code == 202 and "Created JPred job" or "You have successfully submitted" in response.text


@pytest.mark.parametrize("jobid,results_dir_path,extract,silent,host,jpred4", [
    ("jp_TEST", None, False, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4"),
    ("jp_TEST", "jpred_results", False, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4"),
    ("jp_TEST", "jpred_results", True, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4")
])
def test_status(jobid, results_dir_path, extract, silent, host, jpred4):
    with patch("jpredapi.status") as mock_status:
        mock_status.return_value.status_code = 200
        response = jpredapi.status(jobid=jobid, results_dir_path=results_dir_path, extract=extract,
                                   silent=silent, host=host, jpred4=jpred4)
    assert response.status_code == 200


@pytest.mark.skipif(SKIP_REAL_JPREDAPI, reason="Skipping tests that hit the real JPred API server.")
@pytest.mark.parametrize("results_dir_path,extract,silent,host,jpred4", [
    (None, False, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4"),
    ("jpred_results", False, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4"),
    ("jpred_results", True, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4")
])
def test_status_real(results_dir_path, extract, silent, host, jpred4):
    job_response = jpredapi.submit(mode="single", user_format="raw", seq="MQVWPIEGIKKFETLSYLPP")
    result_url = job_response.headers['Location']
    jobid = re.search(r"(jp_.*)$", result_url).group(1)

    response = jpredapi.status(jobid=jobid, results_dir_path=results_dir_path, extract=extract,
                               silent=silent, host=host, jpred4=jpred4)
    assert response.status_code == 200


@pytest.mark.parametrize("jobid,results_dir_path,extract,silent,host,jpred4", [
    ("jp_TEST", None, False, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4"),
    ("jp_TEST", "jpred_results", False, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4"),
    ("jp_TEST", "jpred_results", True, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4")
])
def test_get_results(jobid, results_dir_path, extract, silent, host, jpred4):
    with patch("jpredapi.get_results") as mock_results:
        mock_results.return_value.status_code = 200
        response = jpredapi.status(jobid=jobid, results_dir_path=results_dir_path, extract=extract,
                                   silent=silent, host=host, jpred4=jpred4)
    assert response.status_code == 200


@pytest.mark.skipif(SKIP_REAL_JPREDAPI, reason="Skipping tests that hit the real JPred API server.")
@pytest.mark.parametrize("results_dir_path,extract,silent,host,jpred4", [
    (None, False, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4"),
    ("jpred_results", False, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4"),
    ("jpred_results", True, False, "http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", "http://www.compbio.dundee.ac.uk/jpred4")
])
def test_get_results_real(results_dir_path, extract, silent, host, jpred4):
    job_response = jpredapi.submit(mode="single", user_format="raw", seq="MQVWPIEGIKKFETLSYLPP")
    result_url = job_response.headers['Location']
    jobid = re.search(r"(jp_.*)$", result_url).group(1)

    response = jpredapi.get_results(jobid=jobid, results_dir_path=results_dir_path, extract=extract,
                                    silent=silent, host=host, jpred4=jpred4)
    assert response.status_code == 200


jpredapi/api.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from __future__ import print_function
import re
import os
import tarfile
from collections import OrderedDict

import requests
from retrying import retry


WAIT_INTERVAL = 60000  # 60000 milliseconds = 60 seconds
MAX_ATTEMPTS = 10


def check_rest_version(host="http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", suffix="version", silent=False):
    """Check version of JPred REST interface.

    :param str host: JPred host address.
    :param str suffix: Host address suffix.
    :param silent: Should the work be done silently?
    :type silent: :py:obj:`True` or :py:obj:`False`
    :return: Version of JPred REST API.
    :rtype: :py:class:`str`
    """
    version_url = "{}/{}".format(host, suffix)
    response = requests.get(version_url)
    version = re.search(r"VERSION=(v\.[0-9]*.[0-9]*)", response.text).group(1)
    if not silent:
         print(version)
    return version


def quota(email, host="http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest", suffix="quota", silent=False):
    """Check how many jobs you have already submitted on a given day
    (out of 1000 maximum allowed jobs per user per day).

    :param str email: E-mail address.
    :param str host: JPred host address.
    :param str suffix: Host address suffix.
    :param silent: Should the work be done silently?
    :type silent: :py:obj:`True` or :py:obj:`False`
    :return: Response.
    :rtype: requests.Response
    """
    quota_url = "{}/{}/{}".format(host, suffix, email)
    response = requests.get(quota_url)
    if not silent:
        print(response.text)
    return response


def submit(mode, user_format, file=None, seq=None, skipPDB=True, email=None, name=None, silent=False,
           host="http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest"):
    """Submit job to JPred REST API.

    :param str mode: Submission mode, possible values: `single`, `batch`, `msa`.
    :param str user_format: Submission format, possible values: `raw`, `fasta`, `msf`, `blc`.
    :param str file: File path to a file with the job input (sequence or msa).
    :param str seq: Alternatively, amino acid sequence passed as string of single-letter code without spaces, e.g. --seq=ATWFGTHY
    :param skipPDB: Should the PDB query be skipped?
    :type skipPDB: :py:obj:`True` or :py:obj:`False`
    :param str email: For a batch job submission, where to send the results?
    :param str name: A name for the job.
    :param silent: Should the work be done silently?
    :type silent: :py:obj:`True` or :py:obj:`False`
    :return: Response.
    :rtype: requests.Response
    """
    rest_format = _resolve_rest_format(mode=mode, user_format=user_format)
    query = _create_jpred_query(rest_format=rest_format, file=file, seq=seq,
                                skipPDB=skipPDB, email=email, name=name, silent=silent)

    response = requests.post("{}/{}".format(host, "job"),
                             data=query.encode("utf-8"),
                             headers={"Content-type": "text/txt"})

    if response.status_code == 202 and "created jred job" in response.text.lower():
        if rest_format != "batch":
            result_url = response.headers['Location']
            jobid = re.search(r"(jp_.*)$", result_url).group(1)

            if not silent:
                print("Created JPred job with jobid:", jobid)
                print("You can check the status of the job using the following URL:", result_url)

        elif rest_format == "batch":
            if not silent:
                print(response.text)
    else:
        if not silent:
            print(response.text, response.reason)

    return response


@retry(wait_fixed=WAIT_INTERVAL, stop_max_attempt_number=MAX_ATTEMPTS)
def status(jobid, results_dir_path=None, extract=False, silent=False,
           host="http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest",
           jpred4="http://www.compbio.dundee.ac.uk/jpred4"):
    """Check status of the submitted job.

    :param str jobid: Job id.
    :param str results_dir_path: Directory path where to save results if job is finished.
    :param extract: Extract (True) or not (False) results into directory.
    :type extract: :py:obj:`True` or :py:obj:`False`
    :param silent: Should the work be done silently?
    :type silent: :py:obj:`True` or :py:obj:`False`
    :param str host: JPred host address.
    :param str jpred4: JPred address for results retrieval.
    :return: Response.
    :rtype: requests.Response
    """
    if not silent:
        print("Your job status will be checked with the following parameters:")
        print("Job id:", jobid)
        print("Get results:", bool(results_dir_path))

    job_url = "{}/{}/{}/{}".format(host, "job", "id", jobid)
    response = requests.get(job_url)

    if response.reason == "OK":
        if not silent:
            print(response.text)

        if "finished" in response.text.lower():
            if results_dir_path is not None:
                results_dir_path = os.path.join(results_dir_path, jobid)
                if not os.path.exists(results_dir_path):
                    os.makedirs(results_dir_path)

                archive_url = "{}/{}/{}/{}.{}".format(jpred4, "results", jobid, jobid, "tar.gz")
                archive_path = os.path.join(results_dir_path, "{}.{}".format(jobid, "tar.gz"))

                archive_response = requests.get(archive_url, stream=True)
                with open(archive_path, "wb") as outfile:
                    for chunk in archive_response.iter_content(chunk_size=1024):
                        outfile.write(chunk)

                if extract:
                    tar_archive = tarfile.open(archive_path)
                    tar_archive.extractall(path=results_dir_path)

                if not silent:
                    print("Saving results to: {}".format(os.path.abspath(results_dir_path)))
    else:
        response.raise_for_status()

    return response


def get_results(jobid, results_dir_path=None, extract=False, silent=False,
                host="http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest",
                jpred4="http://www.compbio.dundee.ac.uk/jpred4"):
    """Download results from JPred server.

    :param str jobid: Job id.
    :param str results_dir_path: Directory path where to save results if job is finished.
    :param extract: Extract (True) or not (False) results into directory.
    :type extract: :py:obj:`True` or :py:obj:`False`
    :param silent: Should the work be done silently?
    :type silent: :py:obj:`True` or :py:obj:`False`
    :param str host: JPred host address.
    :param str jpred4: JPred address for results retrieval.
    :return: Response.
    :rtype: requests.Response
    """
    if results_dir_path is None:
        results_dir_path = os.path.join(os.getcwd(), jobid)
    return status(jobid=jobid, results_dir_path=results_dir_path, extract=extract, silent=silent, host=host, jpred4=jpred4)


def _resolve_rest_format(mode, user_format):
    """Resolve format of submission to JPred REST interface based on provided mode and user format.

    :param str mode: Submission mode, possible values: `single`, `batch`, `msa`.
    :param str user_format: Submission format, possible values: `raw`, `fasta`, `msf`, `blc`.
    :return: Format for JPred REST interface.
    :rtype: :py:class:`str`
    """
    if user_format == "raw" and mode == "single":
        rest_format = "seq"
    elif user_format == "fasta" and mode == "single":
        rest_format = "seq"
    elif user_format == "fasta" and mode == "msa":
        rest_format = "fasta"
    elif user_format == "msf" and mode == "msa":
        rest_format = "msf"
    elif user_format == "blc" and mode == "msa":
        rest_format = "blc"
    elif user_format == "fasta" and mode == "batch":
        rest_format = "batch"
    else:
        raise ValueError("""Invalid mode/format combination.
        Valid combinations are: --mode=single --format=raw
                                --mode=single --format=fasta
                                --mode=msa    --format=fasta
                                --mode=msa    --format=msf
                                --mode=msa    --format=blc
                                --mode=batch  --format=fasta""")
    return rest_format


def _create_jpred_query(rest_format, file=None, seq=None, skipPDB=True, email=None, name=None, silent=False):
    """Create query string to be submitted to JPred server.

    :param str rest_format: Format for JPred REST interface.
    :param str file: File path to a file with the job input (sequence or msa).
    :param str seq: Alternatively, amino acid sequence passed as string of single-letter code without spaces, e.g. --seq=ATWFGTHY
    :param skipPDB: Should the PDB query be skipped?
    :type skipPDB: :py:obj:`True` or :py:obj:`False`
    :param str email: For a batch job submission, where to send the results?
    :param str name: A name for the job.
    :param silent: Should the work be done silently?
    :type silent: :py:obj:`True` or :py:obj:`False`
    :return: Query string.
    :rtype: :py:class:`str`
    """
    if file is None and seq is None:
        raise ValueError("""Neither input sequence nor input file are defined.
        Please provide either --file or --seq parameters.""")
    elif file and seq:
        raise ValueError("""Both input sequence and input file are defined.
        Please choose --file or --seq parameter.""")

    if file is not None:
        with open(file, "r") as infile:
            sequence_query = infile.read()
    elif seq is not None:
        sequence_query = ">query\n{}".format(seq)
    else:
        sequence_query = ""

    if skipPDB:
        skipPDB = "on"
    else:
        skipPDB = "off"

    if rest_format == "batch" and email is None:
        raise ValueError("""When submitting batch job email is obligatory.
        You will receive detailed report, list of links and a link to archive
        to all results via email.""")

    if not silent:
        print("Your job will be submitted with the following parameters:")
        print("format:", rest_format)
        print("skipPDB:", skipPDB)
        if file is not None:
            print("file:", file)
        elif seq is not None:
            print("seq:", seq)
        if email is not None:
            print("email:", email)
        if name is not None:
            print("name:", name)

    parameters_dict = OrderedDict([("skipPDB", skipPDB),
                                   ("format", rest_format),
                                   ("email", email),
                                   ("name", name)])

    parameters_list = ["{}={}".format(k, v) for k, v in parameters_dict.items() if v]
    parameters_list.append(sequence_query)
    query = u"£€£€".join(parameters_list)
    return query


jpredapi/__init__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


"""
jpredapi Python library
~~~~~~~~~~~~~~~~~~~~~~~
The JPred API allows users to submit jobs from the command-line.

Usage example for command-line:

   .. code:: bash

      python3 -m jpredapi --help
      python3 -m jpredapi --version
      python3 -m jpredapi submit --mode=single --format=raw --seq=MQVWPIEGIKKFETLSYLPP
      python3 -m jpredapi status --jobid=jp_K46D05A
      python3 -m jpredapi get_results --jobid=jp_K46D05A --results=jpred_sspred/results
      python3 -m jpredapi quota --email=name@domain.com
      python3 -m jpredapi check_rest_version

Usage example for interactive Python shell:

   >>> import jpredapi
   >>>
   >>> jpredapi.submit(mode="single", user_format="raw", seq="MQVWPIEGIKKFETLSYLPP")
   >>>
   >>> jpredapi.status(jobid="jp_K46D05A")
   >>>
   >>> jpredapi.get_results(jobid="jp_K46D05A", results_dir_path="jpred_sspred/results")
   >>>
   >>> jpredapi.quota(email="name@domain.com")
   >>>
   >>> jpredapi.check_rest_version()
   >>>
"""

from .api import submit
from .api import status
from .api import get_results
from .api import quota
from .api import check_rest_version


__version__ = "1.5.6"


jpredapi/cli.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
jpredapi command-line interface

The JPred API allows users to submit jobs from the command-line.

Usage:
    jpredapi submit (--mode=<mode> --format=<format>) (--file=<filename> | --seq=<sequence>) [--email=<name@domain.com>] [--name=<name>] [--rest=<address>] [--skipPDB] [--silent]
    jpredapi status (--jobid=<id>) [--results=<path>] [--wait=<interval>] [--attempts=<max>] [--rest=<address>] [--jpred4=<address>] [--extract] [--silent]
    jpredapi get_results (--jobid=<id>) [--results=<path>] [--wait=<interval>] [--attempts=<max>] [--rest=<address>] [--jpred4=<address>] [--extract] [--silent]
    jpredapi quota (--email=<name@domain.com>) [--rest=<address>] [--silent]
    jpredapi check_rest_version [--rest=<address>] [--silent]
    jpredapi -h | --help
    jpredapi -v | --version

Options:
    -h, --help                 Show this help message.
    -v, --version              Show jpredapi package version.
    --silent                   Do not print messages.
    --extract                  Extract results tar.gz archive.
    --skipPDB                  PDB check.
    --mode=<mode>              Submission mode, possible values: single, batch, msa.
    --format=<format>          Submission format, possible values: raw, fasta, msf, blc.
    --file=<filename>          Filename of a file with the job input (sequence(s)).
    --seq=<sequence>           Instead of passing input file, for single-sequence submission.
    --email=<name@domain.com>  E-mail address where job report will be sent (optional for all but batch submissions).
    --name=<name>              Job name.
    --jobid=<id>               Job id.
    --results=<path>           Path to directory where to save archive with results.
    --rest=<address>           REST address of server [default: http://www.compbio.dundee.ac.uk/jpred4/cgi-bin/rest].
    --jpred4=<address>         Address of Jpred4 server [default: http://www.compbio.dundee.ac.uk/jpred4].
    --wait=<interval>          Wait interval before retrying to check job status in seconds [default: 60].
    --attempts=<max>           Maximum number of attempts to check job status [default: 10].
"""

from . import api


def cli(cmdargs):
    """jpredapi command-line interface processor."""

    api.WAIT_INTERVAL = float(cmdargs["--wait"]) * 1000  # convert from seconds to milliseconds
    api.MAX_ATTEMPTS = int(cmdargs["--attempts"])

    if cmdargs["submit"]:
        api.submit(mode=cmdargs["--mode"],
                   user_format=cmdargs["--format"],
                   file=cmdargs["--file"],
                   seq=cmdargs["--seq"],
                   skipPDB=cmdargs["--skipPDB"],
                   email=cmdargs["--email"],
                   name=cmdargs["--name"],
                   silent=cmdargs["--silent"],
                   host=cmdargs["--rest"])

    elif cmdargs["status"]:
        api.status(jobid=cmdargs["--jobid"],
                   results_dir_path=cmdargs["--results"],
                   extract=cmdargs["--extract"],
                   silent=cmdargs["--silent"],
                   host=cmdargs["--rest"],
                   jpred4=cmdargs["--jpred4"])

    elif cmdargs["get_results"]:
        api.get_results(jobid=cmdargs["--jobid"],
                        results_dir_path=cmdargs["--results"],
                        extract=cmdargs["--extract"],
                        silent=cmdargs["--silent"],
                        host=cmdargs["--rest"],
                        jpred4=cmdargs["--jpred4"])

    elif cmdargs["quota"]:
        api.quota(email=cmdargs["--email"],
                  host=cmdargs["--rest"],
                  suffix="quota",
                  silent=cmdargs["--silent"])

    elif cmdargs["check_rest_version"]:
        api.check_rest_version(host=cmdargs["--rest"],
                               suffix="version",
                               silent=cmdargs["--silent"])
